### PR TITLE
ci: Move to the codecov github action

### PR DESCRIPTION
**Pull request by nirbheek**
The pypi package was suddenly removed. Not the most confidence-inspiring deprecation/migration:

https://about.codecov.io/blog/message-regarding-the-pypi-package/

### CODE REVIEW

**Comment by xclaesse** [comment]
To be honest we could also just remove it completely as far as I'm concern concerned.

**Comment by nirbheek** [comment]
I'm a bit hesitant to do it because people have spent a lot of time adding it. If I can't figure it out, I'll remove it :)

**Comment by xclaesse** [comment]
Let's not fall into the sunk cost fallacy, but if you can make it work it would be nice indeed.

**Comment by nirbheek** [comment]
I'm not saying we should keep it because we've spent a lot of time on it. I'm saying it seems worth keeping it because lots of people have been interested enough in it to contribute it and get it working on all platforms. That's not about the sunk-cost, it's a statement about the usefulness of it to people.

**Comment by eli-schwartz** [comment]
Maybe the codecov action is picky about names and doesn't support ./ for current directory? This shouldn't even ever be necessary for any purposes other than as arguments to a CLI program that parses filenames such as ./--myfile as a command option rather than a filename, but we don't use coverage.xml files with such weird names. :)

**Comment by eli-schwartz** [comment]
I think the reason that the xml file didn't exist is actually because ci/upload_cov.sh is responsible for creating it, turns out:

https://github.com/mesonbuild/meson/blob/8d2a9e6e5c10f5e6c45d7954a063d015cb33bf8c/ci/upload_cov.sh#L1-L13

The action isn't quite the same thing. In fact it prints out about 3.5k lines of... something, e.g. "Found 1739 possible coverage files" followed by one filename per line, our shell script finishes in 250 lines.

**Comment by nirbheek** [comment]
@mensinda @rilian-la-te could you take a look at the LLVM-16 failure with the CMake dependency? 

https://github.com/mesonbuild/meson/actions/runs/4725062379/jobs/8383624738?pr=11682

```
ERR:
CMake Error at /usr/local/opt/llvm/lib/cmake/llvm/LLVM-Config.cmake:134 (if):
  if given arguments:

    "TARGET" "IN_LIST" "LLVM_TARGETS_TO_BUILD"

  Unknown arguments specified
Call Stack (most recent call first):
  /usr/local/opt/llvm/lib/cmake/llvm/LLVM-Config.cmake:244 (llvm_expand_pseudo_components)
  CMakeLists.txt:136 (llvm_map_components_to_libnames)
```

Steps to reproduce:

1. Install llvm 16 (easiest is `brew install llvm` + `export PATH="/usr/local/opt/llvm/bin:$PATH"`)
2. `cd test cases/frameworks/15 llvm && ../../../meson.py builddir -Dmethod=combination -Dlink-static=true`

I tried to look into it, but this CMake macro hackery is beyond me :smile:

**Comment by eli-schwartz** [comment]
We are already tracking this in #11642

**Comment by eli-schwartz** [comment]
There are two failures that are relevant:
UnusedMissingReturn Windows: https://github.com/mesonbuild/meson/actions/runs/4725062389/jobs/8383591133?pr=11682
gccx64ninja (cygwin): https://github.com/mesonbuild/meson/actions/runs/4725062375/jobs/8383003931?pr=11682

Both of these report failure to detect a coverage file for upload. The aggregate step has no output, not even our echo steps. Checking historic CI runs, it never did anything there either, but this went unnoticed because the script that isn't running is the script that tried to upload codecov results. ;)

**Comment by eli-schwartz** [comment]
@nirbheek commit 4cd4f987708d5444a1cf59be14f2774b9f35f940 fixes cygwin

**Comment by nirbheek** [comment]
This should fix windows. For llvm, I am inclined to have the test silently skip for llvm 16 in just the stable branch.

File: .github/workflows/cygwin.yml [hunk 1]
```diff
@@ -102,8 +102,17 @@ jobs:
         # test log should be saved on failure
         if: ${{ !cancelled() }}
 
-      - name: Upload coverage report
+
```
Comment (eli-schwartz) [inline .github/workflows/cygwin.yml:105]: Hmm, I think we probably do not need two lines here.

**Comment by eli-schwartz** [comment]
> [2023-04-20T15:32:44.825Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}

Is this going to happen a lot?

**Comment by nirbheek** [comment]
> Is this going to happen a lot?

I think it happened because I pushed too frequently. If it happens too much we can set `fail_ci_if_error: false`.

**Pull request merged**